pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1757,6 +1758,10 @@ public void deleteOfflineTable(String tableName) {
     HelixHelper.removeResourceFromBrokerIdealState(_helixZkManager, offlineTableName);
     LOGGER.info("Deleting table {}: Removed from broker resource", offlineTableName);
 
+    // Drop the table on servers
+    deleteTableOnServer(offlineTableName);
```

Comment:
As we discussed offline, we need to change the order to avoid some resources not getting cleaned up correctly. 

- delete segments (this gets invoked by cleaning up the idealstate, which is triggered by dropResource() -> please refer PinotHelixResourceManager:deleteOfflineTable()
- wait until the external view is correctly cleaned up
- send the message to drop the table data manger on the server

Reply:
After taking a deeper look, helix messages will be sent to the instances in `Externalview` or `Idealstate`. Helix will throw errors if the `externalview/idealstate` is null. So we have to send the message before cleaning up the `idealstate`. I added the logic on server side to delete the `tableDataManager` only after `externalview `is cleaned up.

Reply:
Helix will throw the exception if `externalview` is null, and the message will not be sent.